README.md
# uae4arm-libretro

Based on old uae4arm version.

## Controls

|RetroPad button|Action|
|---|---|
|B|Fire button 1 / Red|
|A|Fire button 2 / Blue|
|L2|Left mouse button|
|R2|Right mouse button|
|L|Switch to previous disk|
|R|Switch to next disk|
|Select|Toggle virtual keyboard|
|Start|Toggle mouse emulation|

Right analog stick controls the mouse.

In mouse emulation mode dpad and fire button controls the mouse.

Two joysticks support. Switch automatically between mouse or second joystick when a mouse or 2nd joystick button is pressed.

L & R button can change DF0: current disk for multiple disk roms. Each disk should be named with "(Disk x of y)"

Automatic support of WHDload slave in .lha file. 

Kickstarts supported in options configuration:

|System|Version|Filename|Size|MD5|
|---|---|---|---|---|
|A500|KS v1.3 rev 34.005|**kick34005.A500**|262 144|82a21c1890cae844b3df741f2762d48d|
|A600|KS v3.1 rev 40.063|**kick40063.A600**|524 288|e40a5dfb3d017ba8779faba30cbd1c8e|
|A1200|KS v3.1 rev 40.068|**kick40068.A1200**|524 288|646773759326fbac3b2311fd8c8793ee|

If kickstart is not present, AROS kickstart replacement will be used.

changelog.txt
2019-05-31 - Fix atomic functions for aarch64
2019-05-20 - Preparations for aarch64 JIT
2019-05-18 - aarch64: fix 32bpp in Picasso, reduced size of some vars
2019-05-17   First version for aarch64
 - Tested on Raspberry Pi 3B with Ubuntu Mate 64Bit.
 - Compiled native on Raspberry (GCC 7.4.0) or using cross-compiler on Ubuntu PC (GCC 8.3.0).
 - There is still work to do:
    * No JIT (rewrite of code generator needed, because of different instruction set and instruction encoding)
    * Fix for atomic functions (get segfaults when using GCC builtins)
    * Optimize for aarch64
2018-12-03 - JIT: Bugfix CMPA.W (fixes OS3.9 installer)
           - Fix for add folder as HD
2018-10-06 - JIT: Bugfix for issue #1, MVMEL #<data>.W,(An)+
2018-10-04 - JIT: Bugfix ROXR
2018-04-01 - JIT: performance enhancements, implemented ROXR
2018-03-09 - JIT: bugfixes and small enhancements
2018-02-28 - Fixes for non ARMv6T2 target
2018-02-23 - Small enhancements and fixes (for non ARMv6T2)
2018-02-16 - Performance enhancements in JIT and fix for MOVE16 in JIT
2018-02-14 - Implemented ROXL and DIVL (32 bit) for JIT
2018-02-08 - Bugfixes for FPU JIT
2018-01-29 - Added JIT for FPU

2017-11-29 - Added line doubling to remove flicker in interlace mode

2017-11-08 - First version for Raspi with render thread
   - My first version for RaspberryPi using code from Chips and Midwan
   - I introduced new render thread for better performance on hosts with multi core cpus
   
2017-09-09 - Release version
   - Fixed frameskip (was broken in last two beta versions)
   - Fixed Picasso framerate (Pandora switches now to 60 Hz like the Amiga side)
   - Added help page for every panel in GUI (Help button or F1)
   - Fixed bug when emulator crashes or hangs after switching configs

2017-09-06 - Beta version
   - Merged lot of changes from WinUAE 3.5.0
   - Fixed possible segfault and bus error on two locations
   - Better performance by using gcc 7.2 with profiler 

2017-08-08 - Beta version
   - Fixed loading new config (now using same method like WinUAE)
   - Display message instead of crash in some situations
   - More cleanup of threads/pipes/mutex

2017-08-06 - Beta version
   - Merged lot of changes from WinUAE 3.3.0
   - Added panel "Quickstart"
   - Fixed bug when first click in popup window was ignored
   - Support for more compression formats: xz, fdi
   - Support for disk write protect
   - Support for HD write protect
   - Support for Action Replay
   - Support for A600/A1200/A4000 IDE controller
   - Support for Superhires
   - Support for A4000 motherboard/processorboard RAM
   - Support for CD32 fmv module
   - Better handling of cpu fastest possible (less frameskips and less audio hickups)

2016-09-24 - Release version
   - Merged lot of changes from WinUAE 2.8.1
   - Fixed graphic glitches caused by borderblank
   - Fixed bug in navigation through quit button
   - Keys V and W were switched
   - Compiled with gcc 6.2

2016-08-22 - Release version
   - Border sprites (tested with Speris Legacy)
   - CPU fastest: automatic timing adjust to reduce stutter 

2016-08-21 - Beta version
   - Avoid audio delays in some games/demos
   - Automatic switch of freshrate for programmed NTSC/PAL mode change
   - Bugfix in JIT which sometimes caused seg faults when enter gui
   - Savestates enabled for HD and CD configurations (don't know if this is stable)
   - Optimized with profiler
   - Added option "Wait for blits" from WinUAE

2016-07-10 - Beta version
   - JIT: fixed code for non ARMv6T2 CPUs (not relevant for Pandora)
   - Added CD32 buttons to custom control mapping
   - CD32: support for FLAC and MP3 (Audio stutters at start of each audio track because of high cpu usage for decoding. Use frameskip or overclocking to avoid this.)

2016-07-08 - Beta version
   - Performance improvement in JIT
   - Faster HAM decoding
   - Faster sprite drawing in AGA
   - Left + right shoulder + r: switch lores/hires
   - CD32 support
   - New inputdevice handling
   - Built-in configurations for Amiga 500, Amiga 1200 and CD32

2016-05-12 - Release version with new JITv2

2016-05-08 - Beta version
   - Bug fixes in JIT
   - Performance improvement in JIT
   
2016-05-05 - Beta version with new JITv2

2016-04-24 - Release version
   - Improvments in rp9 support (most demos with 68040 CPU now working)
   - Newest code from WinUAE for opcode emulation
   - Minor bugfixes introduced in previous betas 

2016-04-17 - Beta version
   - Bugfix in dialog for adding HDF
   - Improved support for rp9 (only 5 out of 99 games from "Amiga Forever 2016" not working)
     Usage: Copy rp9 files and the three "workbench-xyz.hdf" files to folder "appdata/uae4arm/rp9".
            All files of a rp9 archive added to the last used disk list, so you can easily select them with
            the dropdown control in the floppy panel.
   - Support for RDB hardfiles

2016-04-03 - Beta version
   - Disk images in zip files
   - Support for rp9
   - New mousehack code
   - Picasso96: 8 bit color mode
   - Max. Z3 mem now 128 MB
   - Built-in AROS
   - Bugfix: Press q in a textfield in GUI quits UAE4ARM
   - bsdsocket.library 

2015-10-26 - Release version with bug fixes
   - Switched to newer picasso code from WinUAE. Works much better, even with JIT.
   - Minor update of sprite handling
   - Corrected cpu cycle count in exceptions
   - Minor graphic gliteches fixed (Love-Demo)
   - Compiled with gcc 5.2
  
2015-10-03 - First release
  Changes:
   - Cleanup of code
   - Minor bugfixes
   - Updated CPU cycles for 68020
   - GUI: create hardfile and floppy disks
    
2015-08-16 - Alpha version
  Changes:
   - JIT: minor optimization of emitted ARM code
   - JIT: bugfix for cycle count
   - JIT: bugfix in emitted code (CMP.W, CMP.B, TST.W and TST. B)
   - JIT: disabled option which caused many problems in Picasso96 (JIT now ~15% slower...)
   - Fixed handling of HDs
   - Bugfix: switch from non JIT to JIT now working after first reset
   - DMS file support

src/scsi.cpp
/*
* UAE - The Un*x Amiga Emulator
*
* SCSI and SASI emulation (not uaescsi.device)
*
* Copyright 2007-2015 Toni Wilen
*
*/

#include "sysconfig.h"
#include "sysdeps.h"

#include "options.h"
#include "filesys.h"
#include "blkdev.h"
#include "zfile.h"
#include "memory.h"
#include "scsi.h"
#include "autoconf.h"
#include "rommgr.h"
#include "newcpu.h"
#include "custom.h"
#include "cia.h"

#define NCR5380_SUPRA 1
#define NONCR_GOLEM 2
#define NCR5380_STARDRIVE 3
#define NONCR_KOMMOS 4
#define NONCR_VECTOR 5
#define NONCR_APOLLO 6
#define NCR5380_PROTAR 7
#define NCR5380_ADD500 8
#define NCR5380_KRONOS 9
#define NCR5380_ADSCSI 10
#define NCR5380_ROCHARD 11
#define NCR5380_CLTD 12
#define NCR5380_PTNEXUS 13
#define NCR5380_DATAFLYER 14
#define NONCR_TECMAR 15
#define NCR5380_XEBEC 16
#define NONCR_MICROFORGE 17
#define NONCR_PARADOX 18
#define OMTI_HDA506 19
#define OMTI_ALF1 20
#define OMTI_PROMIGOS 21
#define OMTI_SYSTEM2000 22
#define OMTI_ADAPTER 23
#define OMTI_X86 24
#define NCR5380_PHOENIXBOARD 25
#define NCR5380_TRUMPCARDPRO 26
#define NCR_LAST 27

static const int outcmd[] = { 0x04, 0x0a, 0x0c, 0x2a, 0xaa, 0x15, 0x55, 0x0f, -1 };
static const int incmd[] = { 0x01, 0x03, 0x08, 0x12, 0x1a, 0x5a, 0x25, 0x28, 0x34, 0x37, 0x42, 0x43, 0xa8, 0x51, 0x52, 0xb9, 0xbd, 0xbe, -1 };
static const int nonecmd[] = { 0x00, 0x05, 0x09, 0x0b, 0x11, 0x16, 0x17, 0x19, 0x1b, 0x1e, 0x2b, 0x35, 0xe0, 0xe3, 0xe4, -1 };
static const int scsicmdsizes[] = { 6, 10, 10, 12, 16, 12, 10, 6 };

static void scsi_illegal_command(struct scsi_data *sd)
{
	uae_u8 *s = sd->sense;

	memset (s, 0, sizeof (sd->sense));
	sd->status = SCSI_STATUS_CHECK_CONDITION;
	s[0] = 0x70;
	s[2] = 5; /* ILLEGAL REQUEST */
	s[12] = 0x24; /* ILLEGAL FIELD IN CDB */
	sd->sense_len = 0x12;
}

static void scsi_grow_buffer(struct scsi_data *sd, int newsize)
{
	if (sd->buffer_size >= newsize)
		return;
	uae_u8 *oldbuf = sd->buffer;
	int oldsize = sd->buffer_size;
	sd->buffer_size = newsize + SCSI_DEFAULT_DATA_BUFFER_SIZE;
	write_log(_T("SCSI buffer %d -> %d\n"), oldsize, sd->buffer_size);
	sd->buffer = xmalloc(uae_u8, sd->buffer_size);
	memcpy(sd->buffer, oldbuf, oldsize);
	xfree(oldbuf);
}

static int scsi_data_dir(struct scsi_data *sd)
{
	int i;
	uae_u8 cmd;

	cmd = sd->cmd[0];
	for (i = 0; outcmd[i] >= 0; i++) {
		if (cmd == outcmd[i]) {
			return 1;
		}
	}
	for (i = 0; incmd[i] >= 0; i++) {
		if (cmd == incmd[i]) {
			return -1;
		}
	}
	for (i = 0; nonecmd[i] >= 0; i++) {
		if (cmd == nonecmd[i]) {
			return 0;
		}
	}
	write_log (_T("SCSI command %02X, no direction specified!\n"), sd->cmd[0]);
	return 0;
}

bool scsi_emulate_analyze (struct scsi_data *sd)
{
	int cmd_len, data_len, data_len2, tmp_len;

	data_len = sd->data_len;
	data_len2 = 0;
	cmd_len = scsicmdsizes[sd->cmd[0] >> 5];
	if (sd->hdhfd && sd->hdhfd->ansi_version < 2 && cmd_len > 10)
		goto nocmd;
	sd->cmd_len = cmd_len;
	switch (sd->cmd[0])
	{
	case 0x04: // FORMAT UNIT
		if (sd->device_type == UAEDEV_CD)
			goto nocmd;
		// FmtData set?
		if (sd->cmd[1] & 0x10) {
			int cl = (sd->cmd[1] & 8) != 0;
			int dlf = sd->cmd[1] & 7;
			data_len2 = 4;
		} else {
			sd->direction = 0;
			sd->data_len = 0;
			return true;
		}
	break;
	case 0x0c: // INITIALIZE DRIVE CHARACTERICS (SASI)
		if (sd->hfd && sd->hfd->ci.unit_feature_level < HD_LEVEL_SASI)
			goto nocmd;
		data_len = 8;
	break;
	case 0x08: // READ(6)
		data_len2 = sd->cmd[4] * sd->blocksize;
		scsi_grow_buffer(sd, data_len2);
	break;
	case 0x28: // READ(10)
		data_len2 = ((sd->cmd[7] << 8) | (sd->cmd[8] << 0)) * (uae_s64)sd->blocksize;
		scsi_grow_buffer(sd, data_len2);
	break;
	case 0xa8: // READ(12)
		data_len2 = ((sd->cmd[6] << 24) | (sd->cmd[7] << 16) | (sd->cmd[8] << 8) | (sd->cmd[9] << 0)) * (uae_s64)sd->blocksize;
		scsi_grow_buffer(sd, data_len2);
	break;
	case 0x0f: // WRITE SECTOR BUFFER
		data_len = sd->blocksize;
		scsi_grow_buffer(sd, data_len);
	break;
	case 0x0a: // WRITE(6)
		if (sd->device_type == UAEDEV_CD)
			goto nocmd;
		data_len = sd->cmd[4] * sd->blocksize;
		scsi_grow_buffer(sd, data_len);
	break;
	case 0x2a: // WRITE(10)
		if (sd->device_type == UAEDEV_CD)
			goto nocmd;
		data_len = ((sd->cmd[7] << 8) | (sd->cmd[8] << 0)) * (uae_s64)sd->blocksize;
		scsi_grow_buffer(sd, data_len);
	break;
	case 0xaa: // WRITE(12)
		if (sd->device_type == UAEDEV_CD)
			goto nocmd;
		data_len = ((sd->cmd[6] << 24) | (sd->cmd[7] << 16) | (sd->cmd[8] << 8) | (sd->cmd[9] << 0)) * (uae_s64)sd->blocksize;
		scsi_grow_buffer(sd, data_len);
	break;
	case 0xbe: // READ CD
	case 0xb9: // READ CD MSF
		if (sd->device_type != UAEDEV_CD)
			goto nocmd;
		tmp_len = (sd->cmd[6] << 16) | (sd->cmd[7] << 8) | sd->cmd[8];
		// max block transfer size, it is usually smaller.
		tmp_len *= 2352 + 96;
		scsi_grow_buffer(sd, tmp_len);
	break;
	case 0x2f: // VERIFY
		if (sd->cmd[1] & 2) {
			sd->data_len = ((sd->cmd[7] << 8) | (sd->cmd[8] << 0)) * (uae_s64)sd->blocksize;
			scsi_grow_buffer(sd, sd->data_len);
			sd->direction = 1;
		} else {
			sd->data_len = 0;
			sd->direction = 0;
		}
		return true;
	case 0x15: // MODE SELECT
	case 0x55: 
	if (sd->device_type != UAEDEV_CD)
		goto nocmd;
	break;
	}
	if (data_len < 0) {
		if (cmd_len == 6)
			sd->data_len = sd->cmd[4];
		else
			sd->data_len = (sd->cmd[7] << 8) | sd->cmd[8];
	} else {
		sd->data_len = data_len;
	}
	sd->direction = scsi_data_dir (sd);
	return true;
nocmd:
	sd->status = SCSI_STATUS_CHECK_CONDITION;
	sd->direction = 0;
	scsi_illegal_command(sd);
	return false;
}

static void scsi_clear_sense(struct scsi_data *sd)
{
	memset (sd->sense, 0, sizeof (sd->sense));
	memset (sd->reply, 0, sizeof (sd->reply));
	sd->sense[0] = 0x70;
}
static void copysense(struct scsi_data *sd)
{
	int len = sd->cmd[4];
	if (len == 0)
		len = 4;
	memset(sd->buffer, 0, len);
	memcpy(sd->buffer, sd->sense, sd->sense_len > len ? len : sd->sense_len);
	if (len > 7 && sd->sense_len > 7)
		sd->buffer[7] = sd->sense_len - 8;
	if (sd->sense_len == 0)
		sd->buffer[0] = 0x70;
	sd->data_len = len;
	scsi_clear_sense(sd);
}
static void copyreply(struct scsi_data *sd)
{
	if (sd->status == 0 && sd->reply_len > 0) {
		memset(sd->buffer, 0, 256);
		memcpy(sd->buffer, sd->reply, sd->reply_len);
		sd->data_len = sd->reply_len;
	}
}

static bool handle_ca(struct scsi_data *sd)
{
	bool cc = sd->sense_len > 2 && sd->sense[2] >= 2;
	bool ua = sd->unit_attention != 0;
	uae_u8 cmd = sd->cmd[0];

	// INQUIRY
	if (cmd == 0x12) {
		if (ua && cc && sd->sense[2] == 6) {
			// INQUIRY clears UA only if previous
			// command was aborted due to UA
			sd->unit_attention = 0;
		}
		memset(sd->reply, 0, sizeof(sd->reply));
		return true;
	}

	// REQUEST SENSE
	if (cmd == 0x03) {
		if (ua) {
			uae_u8 *s = sd->sense;
			scsi_clear_sense(sd);
			s[0] = 0x70;
			s[2] = 6; /* UNIT ATTENTION */
			s[12] = (sd->unit_attention >> 8) & 0xff;
			s[13] = (sd->unit_attention >> 0) & 0xff;
			sd->sense_len = 0x12;
		}
		sd->unit_attention = 0;
		return true;
	}

	scsi_clear_sense(sd);

	if (ua) {
		uae_u8 *s = sd->sense;
		s[0] = 0x70;
		s[2] = 6; /* UNIT ATTENTION */
		s[12] = (sd->unit_attention >> 8) & 0xff;
		s[13] = (sd->unit_attention >> 0) & 0xff;
		sd->sense_len = 0x12;
		sd->unit_attention = 0;
		sd->status = 2;
		return false;
	}

	return true;
}

void scsi_emulate_cmd(struct scsi_data *sd)
{
	sd->status = 0;
	if ((sd->message[0] & 0xc0) == 0x80 && (sd->message[0] & 0x1f)) {
		uae_u8 lun = sd->message[0] & 0x1f;
		if (lun > 7)
			lun = 7;
		sd->cmd[1] &= ~(7 << 5);
		sd->cmd[1] |= lun << 5;
	}
	if (sd->device_type == UAEDEV_CD && sd->cd_emu_unit >= 0) {
		uae_u32 ua = 0;
		ua = scsi_cd_emulate(sd->cd_emu_unit, NULL, 0, 0, 0, 0, 0, 0, 0, sd->atapi);
		if (ua)
			sd->unit_attention = ua;
		if (handle_ca(sd)) {
			if (sd->cmd[0] == 0x03) { /* REQUEST SENSE */
				scsi_cd_emulate(sd->cd_emu_unit, sd->cmd, 0, 0, 0, 0, 0, 0, 0, sd->atapi); /* ack request sense */
				copysense(sd);
			} else {
				sd->status = scsi_cd_emulate(sd->cd_emu_unit, sd->cmd, sd->cmd_len, sd->buffer, &sd->data_len, sd->reply, &sd->reply_len, sd->sense, &sd->sense_len, sd->atapi);
				copyreply(sd);
			}
		}
	} else if (sd->device_type == UAEDEV_HDF) {
		uae_u32 ua = 0;
		ua = scsi_hd_emulate(sd->hfd, sd->hdhfd, NULL, 0, 0, 0, 0, 0, 0, 0);
		if (ua)
			sd->unit_attention = ua;
		if (handle_ca(sd)) {
			if (sd->cmd[0] == 0x03) { /* REQUEST SENSE */
				scsi_hd_emulate(sd->hfd, sd->hdhfd, sd->cmd, 0, 0, 0, 0, 0, sd->sense, &sd->sense_len);
				copysense(sd);
			} else {
				sd->status = scsi_hd_emulate(sd->hfd, sd->hdhfd,
					sd->cmd, sd->cmd_len, sd->buffer, &sd->data_len, sd->reply, &sd->reply_len, sd->sense, &sd->sense_len);
				copyreply(sd);
			}
		}
	}
	sd->offset = 0;
}

static void allocscsibuf(struct scsi_data *sd)
{
	sd->buffer_size = SCSI_DEFAULT_DATA_BUFFER_SIZE;
	sd->buffer = xcalloc(uae_u8, sd->buffer_size);
}

struct scsi_data *scsi_alloc_generic(struct hardfiledata *hfd, int type)
{
	struct scsi_data *sd = xcalloc(struct scsi_data, 1);
	sd->hfd = hfd;
	sd->id = -1;
	sd->cd_emu_unit = -1;
	sd->blocksize = hfd->ci.blocksize;
	sd->device_type = type;
	allocscsibuf(sd);
	return sd;
}

struct scsi_data *scsi_alloc_cd(int id, int unitnum, bool atapi)
{
	struct scsi_data *sd;
	if (!sys_command_open (unitnum)) {
		write_log (_T("SCSI: CD EMU scsi unit %d failed to open\n"), unitnum);
		return NULL;
	}
	sd = xcalloc (struct scsi_data, 1);
	sd->id = id;
	sd->cd_emu_unit = unitnum;
	sd->atapi = atapi;
	sd->blocksize = 2048;
	sd->device_type = UAEDEV_CD;
	allocscsibuf(sd);
	return sd;
}

void scsi_free(struct scsi_data *sd)
{
	if (!sd)
		return;
	if (sd->cd_emu_unit >= 0) {
		sys_command_close (sd->cd_emu_unit);
		sd->cd_emu_unit = -1;
	}
	xfree(sd->buffer);
	xfree(sd);
}

void scsi_start_transfer(struct scsi_data *sd)
{
	sd->offset = 0;
}
